Add upgrade guide for v3 and v4

kmaehashi opened the pull request:
This fixes #3592 and #4143.

I renamed existing upgrade guide (v1 to v2) as `upgrade_v2.rst`, as it is huge.
Then new upgrade guide (v2 to v3, v3 to v4) is placed as `upgrade.rst`, with a link to `upgrade_v2.rst`.
So, I recommend to review per-commit rather than full diff.

When backporting, please remove the v4 part.

niboshi commented:
This PR solves #3592.

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -1,819 +1,124 @@
-   Consider the following model definition and the code to call it in test mode written for Chainer v1.
+Prior to v4, the above code raises an exception like ``ValueError: object __array__ method not producing an array``, which was difficult to understand.
+In v4, the error message would become ``ValueError: numpy and cupy arrays are mixed in the forward input (Maximum)``.
+This kind of error usually occur by mistake (for example, not performing ``to_gpu`` for some variables).
```
`occurs`?

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -1,819 +1,124 @@
 
-      class MyModel(chainer.Link):
-          ...
+Note that old-style function classes (inheriting :class:`Function`) are not affected.
```
This sentence is not true. `FunctionAdapter` raises the error.

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -1,819 +1,124 @@
 
-          def __call__(self, x, train=True):
-              return f(F.dropout(x, train=train))
+Reproducibility Improvement with MultiProcessIterator
```
Do we need this topic in upgrade guide?

kmaehashi replied:
I wrote it here because the issue had `no-compat` label, but I now think it is not needed to be on the upgrade guide, because we don't guarantee that the reproducibility of computation results between any two version.

niboshi replied:
OK. More precisely, in general we don't guarrantee the reproducibility within the same version, if my understanding is correct. (#3575 seems to be an exception but it's not documented anyway).

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -1,819 +1,124 @@
+See the discussion in `#3575 <https://github.com/chainer/chainer/pull/3575>`_ for more details.
 
-   In Chainer v2, it should be updated into the following code:
+Update of Docker Images
```
ditto. (Do we need this topic in upgrade guide?)

kmaehashi replied:
I think this is needed, because CUDA 7.5 users can no longer run the official docker image.

kmaehashi replied:
I mean, users with older NVIDIA driver version cannot run the Docker image, although we still continue to support CUDA 7.5.
https://github.com/NVIDIA/nvidia-docker/wiki/CUDA#requirements

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -1,819 +1,124 @@
+See the `Release Note for v3.0.0 <https://github.com/chainer/chainer/releases/tag/v3.0.0>`_ for the usage of this feature.
 
-.. _upgrade-configurations:
+Many of :doc:`reference/functions` are already migrated to new-style, although some of functions are still old-style (classes inheriting from :class:`Function`).
```
Do we need this paragraph?
Instead, we should state that writing new functions with `FunctionNode` is encouraged.

kmaehashi replied:
In this paragraph I mean "currently not all functions supports double backprop". I'll make it clear.

> we should state that writing new functions with FunctionNode is encouraged.

I agree with this.

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -1,819 +1,124 @@
-- Simply subsituting a new value to an entry, like ``chainer.config.train = False``.
-- Using the ``chainer.using_config`` context manager.
-  It can be used with the ``with`` statement of Python as follows::
+The behavior of :func:`chainer.functions.matmul` has been changed to behave like ``matmul`` function of NumPy.
```
``` ``matmul`` ``` -> Link to `numpy.matmul` documentation

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -1,819 +1,124 @@
 
-Volatile flag is removed
-~~~~~~~~~~~~~~~~~~~~~~~~
+   Codes using ``use_cudnn`` argument of :func:`chainer.functions.spatial_transformer_grid` and :func:`chainer.functions.spatial_transformer_sampler` require modification to work with Chainer v3.
```
codes -> the existing code?

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -1,819 +1,124 @@
+.. attention::
 
-  It recovers the original configuration after quitting the ``with`` block.
+   Codes using :func:`chainer.functions.matmul` may require modification to work with Chainer v3.
```
ditto. (codes -> the existing code?)

niboshi commented:
I think we should mention optimizer setup (#3488, #3940), too.

niboshi commented:
Let's keep from mentioning optimizer setup until the discussioin in  #4141 will be settled.

kmaehashi commented:
Thanks for review! I fixed accordingly and rebased.

> Let's keep from mentioning optimizer setup until the discussioin in #4141 will be settled.

OK.

niboshi commented:
Now #4141 has been merged.
I think it's ok whether or not to mention it, as the difference (between v3 method and meregd one) is not so large.

kmaehashi commented:
#4141 is just a syntax sugar so I feel it is not needed to be in this guide.

(BTW #4141 lacks pydoc fixes ("Returns:") so we should fix that)

niboshi commented:
jenkins, test this please

niboshi commented:
LGTM!
